File: vdf/mimc/python/mimc.py
import time

modulus = 2**256 - 2**32 * 351 + 1
little_fermat_expt = (modulus*2-1)//3
round_constants = [(i**7) ^ 42 for i in range(64)]
input = 3
steps = 8192

# Forward MiMC
start_time = time.time()
inp = input
for i in range(1,steps):
    inp = (inp**3 + round_constants[i % len(round_constants)]) % modulus
print("forward-mimc %.4f sec" % (time.time() - start_time))
output = inp

# Reverse MIMC
rtrace = output
start_time = time.time()
for i in range(steps - 1, 0, -1):
    rtrace = pow(rtrace-round_constants[i%len(round_constants)], little_fermat_expt, modulus)
print("reverse-mimc: %.4f sec" % (time.time() - start_time))

if rtrace == input:
    print("PASS")
else:
    print("FAIL")

File: vdf/mimc/js/mimc.js
var bigInt = require("big-integer");

// 64 round_constants
SEVEN = bigInt(7);
FORTYTWO = bigInt(42);
len_round_constants = 64;
round_constants = new Array();
for (i = 0; i < 64; i++) {
  round_constants[i] = bigInt(i).pow(SEVEN).xor(FORTYTWO);
}

modulus = bigInt(2).pow(256).subtract(bigInt(2).pow(32).multiply(351)).add(1);
nsteps = 8192;

// Forward MiMC
THREE = bigInt(3);
input = bigInt(3);
trace = bigInt(input);
console.time("forward-mimc");
for (i = 1; i < nsteps; i++) {
  trace = trace.pow(THREE).add(round_constants[i%len_round_constants]).mod(modulus);
}
console.timeEnd("forward-mimc");
output = bigInt(trace)

// Reverse MiMC
console.time("reverse-mimc");
rtrace = bigInt(output)
little_fermat_expt = bigInt(modulus).multiply(2).subtract(1).divide(3)
for (i = nsteps - 1; i > 0; i--) {
  rtrace = rtrace.subtract(round_constants[i%len_round_constants]).modPow(little_fermat_expt, modulus);
}
console.timeEnd("reverse-mimc");

// print if input matches
if ( rtrace.equals(input) ) {
  console.log("PASS");
} else {
  console.log("FAIL");
}
